src/core/eme/create_session.ts

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Observable } from "rxjs/Observable";
+import {
+  IMediaKeySession,
+} from "../../compat";
+import { ErrorCodes } from "../../errors";
+import arrayIncludes from "../../utils/array-includes";
+import castToObservable from "../../utils/castToObservable";
+import log from "../../utils/log";
+import {
+  $loadedSessions,
+  $storedSessions,
+  IMediaKeysInfos,
+} from "./constants";
+
+export interface ICreatedSessionEvent {
+  type : "created-session";
+  value : {
+    session : MediaKeySession|IMediaKeySession;
+  };
+}
+
+export interface IPersistentSessionRecoveryEvent {
+  type : "loaded-persistent-session";
+  value : {
+    session : MediaKeySession|IMediaKeySession;
+  };
+}
+
+export type IGetSessionEvent =
+  ICreatedSessionEvent |
+  IPersistentSessionRecoveryEvent;
+
+/**
+ * If session creating fails, retry once session creation/loading.
+ * @param initData
+ * @param initDataType
+ * @param mediaKeysInfos
+ * @returns {Observable}
+ */
+function loadPersistentSession(
+  storedSessionId: string,
+  initData: Uint8Array,
+  initDataType: string,
+  session: MediaKeySession|IMediaKeySession
+) : Observable<IGetSessionEvent> {
+  log.debug("eme: load persisted session", storedSessionId);
+
+  return castToObservable(session.load(storedSessionId))
+    .catch((error) => {
+      log.warn("eme: no data stored for the loaded session.",
+        storedSessionId);
+
+      const loadedSession = $loadedSessions.get(initData, initDataType);
+      if (loadedSession != null) {
+        $loadedSessions.closeSession(loadedSession);
+      }
+      $storedSessions.delete(initData);
+
+      throw error;
+    })
+    .map((success) => {
+      if (success) {
+        $loadedSessions.add(initData, initDataType, session);
+        return {
+          type: "loaded-persistent-session" as "loaded-persistent-session",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=187&issues=AWMmgcqUmtQEh111i0u2&open=AWMmgcqUmtQEh111i0u2)

[ISSUE_KEY]: AWMmgcqUmtQEh111i0u2